src/api.ts

```diff
@@ -0,0 +1,216 @@
+}
+
+type Location = {
+    line: number,
+    column: number,
+}
+
+type GraphQLError = {
+    message: string,
+    path: Array<string>,
+    locations: Array<Location>,
+    extensions?: { [key: string]: string }
+} // this will be from `graphql` package only the type
+
+type GraphqlResult<T extends JSONObject = JSONObject> = {
+    data?: T,
+    errors?: Array<GraphQLError>,
+    extensions?: { [key: string]: unknown }
+}
+
+declare class API {
+    /** HTTP GET METHOD */
+    static get<ResponseBody extends unknown>(params: APIParam, init?: FetchOptions): Promise<APIResponse<ResponseBody>>;
+    /** HTTP HEAD METHOD */
+    static head<ResponseBody extends unknown>(params: APIParam, init?: FetchOptions): Promise<APIResponse<ResponseBody>>;
+    /** HTTP DELETE METHOD */
+    static delete<ResponseBody extends unknown>(params: APIParam, init?: FetchOptions): Promise<APIResponse<ResponseBody>>;
+    /** HTTP OPTIONS METHOD */
+    static options<ResponseBody extends unknown>(params: APIParam, init?: FetchOptions): Promise<APIResponse<ResponseBody>>;
+    /** HTTP PUT METHOD */
+    static put<RequestBody extends JSONValue = JSONValue, ResponseBody extends unknown = unknown>(params: APIParam, init: UpdateOptionsJSON<RequestBody>): Promise<APIResponse<ResponseBody>>
+    static put<RequestBody extends string = string, ResponseBody extends unknown = unknown>(params: APIParam, init: UpdateOptionsString<RequestBody>): Promise<APIResponse<ResponseBody>>
+    /** HTTP POST METHOD */
+    static post<RequestBody extends JSONValue = JSONValue, ResponseBody extends unknown = unknown>(params: APIParam, init: UpdateOptionsJSON<RequestBody>): Promise<APIResponse<ResponseBody>>
+    static post<RequestBody extends string = string, ResponseBody extends unknown = unknown>(params: APIParam, init: UpdateOptionsString<RequestBody>): Promise<APIResponse<ResponseBody>>
+    /** HTTP PATCH METHOD */
+    static patch<RequestBody extends JSONValue = JSONValue, ResponseBody extends unknown = unknown>(params: APIParam, init: UpdateOptionsJSON<RequestBody>): Promise<APIResponse<ResponseBody>>
+    static patch<RequestBody extends string = string, ResponseBody extends unknown = unknown>(params: APIParam, init: UpdateOptionsString<RequestBody>): Promise<APIResponse<ResponseBody>>
+    /** Cancel HTTP Request */
+    static cancel: (request: Promise<unknown>, message: string) => void
+    /** */
+    static query: <GraphqlData extends JSONObject = JSONObject>({ document, variables }: { document: string, variables?: JSONObject }) => Promise<GraphqlResult<GraphqlData>>
+    static mutation: <GraphqlData extends JSONObject = JSONObject>({ document, variables }: { document: string, variables?: JSONObject }) => Promise<GraphqlResult<GraphqlData>>
+
+    static subscription: <GraphqlData extends JSONObject = JSONObject>({ document, variables }: { document: string, variables?: JSONObject }) => Observable<GraphqlResult<GraphqlData>>;
+}
+
+type Observable<T = object> = {
```

Comment:
`object` in JS  can be interpreted as different DS https://www.w3schools.com/js/js_object_definition.asp . We can narrow down the type to Record<string, unknown>

Reply:
Yes! good catch

Reply:
I change it on the `query`, `mutation`, and `subscription` the other is something more generic just to mimic what is doing `Observable` type from `zen-observable-ts`